File: data_structure/LinkedList/python/doubly_linked_list.py
class Node:
    def __init__(self, data, next = None, prev = None):
        self.data = data
        self.next = next
        self.prev = prev


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.count = 0

    def print_all(self):
        current_node = self.head
        text = "["

        for i in range(self.count):
            text += str(current_node.data)
            current_node = current_node.next

            if i is not self.count - 1:
                text += ", "

        text += "]"
        print(text)

    def clear(self):
        self.head = None
        self.count = 0

    def insert_at(self, index, data):
        if index > self.count or index < 0:
            raise IndexError

        new_node = Node(data)
        if index == 0:
            new_node.next = self.head
            if self.head is not None:
                self.head.prev = new_node
            self.head = new_node
        elif index == self.count:
            new_node.next = None
            new_node.prev = self.tail
            self.tail.next = new_node
        else:
            current_node = self.head

            for i in range(index - 1):
                current_node = current_node.next
            new_node.next = current_node.next
            new_node.prev = current_node
            current_node.next = new_node
            new_node.next.prev = new_node

        if new_node.next == None:
            self.tail = new_node

        self.count += 1

    def insert_last(self, data):
        self.insert_at(self.count, data)

    def delete_at(self, index):
        if index >= self.count or index < 0:
            raise IndexError

        current_node = self.head

        if index == 0:
            delete_node = self.head
            if self.head.next is None:
                self.head = None
                self.tail = None
            else:
                self.head = self.head.next
                self.head.prev = None

            self.count -= 1

            return delete_node
        elif index == self.count - 1:
            delete_node = self.tail
            self.tail.prev.next = None
            self.tail = self.tail.prev
            self.count -= 1

            return delete_node
        else:
            for i in range(index - 1):
                current_node = current_node.next

            delete_node = current_node.next
            current_node.next = current_node.next.next
            current_node.next.prev = current_node
            self.count -= 1

            return delete_node

    def delete_last(self):
        return self.delete_at(self.count - 1)

    def get_node_at(self, index):
        if index >= self.count or index < 0:
            raise IndexError("범위를 넘어갔습니다.")

        current_node = self.head
        for i in range(index):
            current_node = current_node.next

        return current_node

File: theory/data_structure/LinkedList/java/doubly_linked_list/DoublyLinkedList.java
package data_structure.linked_list.doubly_linked_list;

public class DoublyLinkedList <T> {
    private Node<T> head;
    private Node<T> tail;
    private int count;

    public DoublyLinkedList() {
        this.head = null;
        this.tail = null;
        this.count = 0;
    }

    /**
     * 전체 요소 출력
     */
    public void printAll() {
        Node<T> currentNode = this.head;
        String text = "[";

        while (currentNode != null) {
            text += currentNode.getData();
            currentNode = currentNode.getNext();

            if (currentNode != null) {
                text += ", ";
            }
        }

        text += "]";
        System.out.println(text);
    }

    /**
     * 전체 요소 초기화
     */
    public void clear() {
        head = null;
        count = 0;
    }

    /**
     * 요소 추가
     */
    public void insertAt(int index, T data) {
        if (index > count || index < 0) {
            throw new IndexOutOfBoundsException("범위를 넘어갔습니다.");
        }

        Node<T> newNode = new Node<>(data);

        if (index == 0) {
            newNode.setNext(head);
            if (head != null) {
                head.setPrev(newNode);
            }
            head = newNode;
        } else if (index == count) {
            newNode.setNext(null);
            newNode.setPrev(tail);
            tail.setNext(newNode);
        } else {
            Node<T> currentNode = head;

            for (int i = 0; i < index - 1; i++) {
                currentNode = currentNode.getNext();
            }
            newNode.setNext(currentNode.getNext());
            newNode.setPrev(currentNode);
            currentNode.setNext(newNode);
            newNode.getNext().setPrev(newNode);
        }

        if (newNode.getNext() == null) {
            tail = newNode;
        }
        count++;
    }

    /**
     * 리스트 끝에 요소 추가
     */
    public void insertLast(T data) {
        insertAt(count, data);
    }

    /**
     * 요소 삭제
     */
    public Node<T> deleteAt(int index) {
        if (index >= this.count || index < 0) {
            throw new IndexOutOfBoundsException("잘못된 범위입니다.");
        }

        Node<T> currentNode = head;

        if (index == 0) {
            Node<T> deleteNode = head;
            if (head.getNext() == null) {
                head = null;
                tail = null;
            } else {
                head = head.getNext();
                head.setPrev(null);
            }

            count--;
            return deleteNode;
        } else if (index == count - 1) {
            Node<T> deleteNode = tail;
            tail.getPrev().setNext(null);
            tail = tail.getPrev();
            count--;

            return deleteNode;
        } else {
            for (int i = 0; i < index - 1; i++) {
                currentNode = currentNode.getNext();
            }

            Node<T> deleteNode = currentNode.getNext();
            currentNode.setNext(currentNode.getNext().getNext());
            currentNode.getNext().setPrev(currentNode);
            count--;

            return deleteNode;
        }
    }

    /**
     * 끝 요소 삭제
     */
    public Node<T> deleteLast() {
        return deleteAt(count - 1);
    }

    /**
     * 요소 조회
     */
    public Node<T> getNodeAt(int index) {
        if (index >= count || index < 0) {
            throw new IndexOutOfBoundsException("잘못된 범위입니다!");
        }

        Node<T> currentNode = head;
        for (int i = 0; i < index; i++) {
            currentNode = currentNode.getNext();
        }

        return currentNode;
    }

    public Node getHead() {
        return head;
    }

    public void setHead(Node head) {
        this.head = head;
    }

    public Node getTail() {
        return tail;
    }

    public void setTail(Node tail) {
        this.tail = tail;
    }

    public int getCount() {
        return count;
    }

    public void setCount(int count) {
        this.count = count;
    }
}
